File: Ocena 5 Marciniak Daniel/zad2.ts
// Zadanie 2a: Pierwiastek sześcienny
const cubeRoot = (x: number, guess: number = x / 3): number => {
    const nextGuess = (g: number): number => (2 * g + x / (g * g)) / 3;
    return Math.abs(guess ** 3 - x) < 1e-10 ? guess : cubeRoot(x, nextGuess((guess)));
};

// Zadanie 2b: Metoda Herona uzależniona od Epsilon
const sqrtEpsilon = (x: number, epsilon: number, guess: number = x / 2): number => {
    const nextGuess = (g: number): number => (g + x / g) / 2;
    return Math.abs(guess ** 2 - x) < epsilon ? guess : sqrtEpsilon(x, epsilon, nextGuess(guess));
};

// Zadanie 2c: Metoda Herona uzależniona od N (ilości kroków)
const sqrtSteps = (x: number, steps: number, guess: number = x / 2): number => {
    const nextGuess = (g: number): number => (g + x / g) / 2;
    return steps === 0 ? guess : sqrtSteps(x, steps - 1, nextGuess(guess));
};

const x = 27; // Vars can be in a parameters in functions e.g. cubeRoot(27), but vars are const

console.log(`Cube root of ${x} is: ${cubeRoot(x)}`); 

const epsilon = 1e-10;
console.log(`Square root of ${x} with epsilon ${epsilon} is: ${sqrtEpsilon(x, epsilon)}`);

const steps = 10;
console.log(`Square root of ${x} with ${steps} steps is: ${sqrtSteps(x, steps)}`);

File: Ocena 3.5 Kasprzak Sebastian/Zad5/Zad2.java
package prod.func.Zad5;

public class Zad2 {
    public static void main(String[] args) {
        System.out.println("Cube root of 27 with Newton's method is: " + cubeRoot(27, 0.0001));
    }

    public static double cubeRoot(double x, double epsilon) {
        return cubeRootRecursive(x, epsilon, x / 3.0);
    }

    private static double cubeRootRecursive(double x, double epsilon, double guess) {
        return ((((guess * guess * guess - x) < 0.0)
            ? -(guess * guess * guess - x)
            : (guess * guess * guess - x) ) < epsilon)
                ? guess
                : cubeRootRecursive(x, epsilon, (2 * guess + x / (guess * guess)) / 3.0);
    }
}
